### ISSUE TITLE
RFC: [Design] Separate SuiDataStore into Modular Components

### ISSUE THREAD

**Issue by velvia** [MystenLabs/sui#891]
(Credit to #629 as well as @gdanezis 's latest PR which have similar themes)

Today storage is tied very intimately into the authority.  Multiple places throughout the transaction flow directly read and write from the database.   This is a design for splitting out the `SuiDataStore` into multiple components which could potentially be offloaded from the authority/validator main process, offloading and cold storage of older or infrequently accessed data, and serving out reads to third parties.  

Such storage components could, in theory, be a completely separate process and be used by the gateway as well.

```mermaid
graph TD
  Authority --> AC(AuthorityCache)
  AC -->|Objects TX| SC(Storage Component)
  Authority --> LS(LockService)
  LS --> LSDB[(LocksDataBase)]
  SC --> TxDB[(TransactionsDatabase)]
  SC --> ODB[(ObjectsDatabase)]
  TxDB --> ColdStorage
  ODB --> ColdStorage
  SC --> AC
```

## Goals

In the above design, the authority's functionality would be simplified to coordinating transactions and shared objects.

- Offload storage functionality from core authority (or gateway)
- Streamline performance in authority by separating most I/O and minimizing locking paths
- Allow for possibilities to experiment with higher-latency, distributed storage
- or with storage that requires reorganization or other optimization of the data (eg schema-aware  or columnar or graph storage)
- Potentially offload third party reads of such data
- Allow for implementation of state shedding / tree pruning and other storage maintenance functionality essential for scalability
- Allow for scalable storage beyond one node

## AuthorityCache

Authorities (Gateways?) no longer talks directly to an AuthorityStore.  Instead, they talk to an `AuthorityCache`, which is a write-through caching layer.
- Since in this design it could take a while for writes to finally be persisted, especially in case of a remote storage layer, the cache allows future transactions to read from the cache before writes have reached final destination, speeding up transaction processing.
- Taking advantage of this requires that subsequent TX's depending on the same object arrives at the same authority
- More conservative stateless mode means TX would have to wait for persistent storage to actually get the new objects

## Changes to the transaction processing flow

Non shared objects (single writer)

1. Authority acquires locks from LockService 
2. Authority writes signed transaction to WAL, returns signed tx to client
3. Cert processing: MOVE processing etc.
4. Writes a new `TransactionChangeLog` to the `AuthorityCache`
5. `AuthorityCache` writes change log to WAL (single writer) and waits for flush (or not, if you want to live faster and more dangerously)
6. Lock service creates new locks for new objects (actually is this necessary?  If locks for existing objects don't exist, then first getter wins automatically right?)

Note that mutexes aren't necessary because the WAL linearizes final writes.

TODO: Shared Object flow

## LockService

The lock service needs to atomically and consistently be able to acquire and clear locks on multiple objects, and its persistence needs to be fast or immediate.  This is to ensure that when we sign a transaction and return it from an authority, the client can have a guarantee that it will have the right, given certificate, to process that transaction, in the event of failure.
My thoughts on implementation:
- Independent RocksDB instance just for locks
- Single threaded loop with MPSC lock free queue to minimize contention
- Immediate persistence
- Commands:
  - Set lock for TX for object-list
  - Check locks for object-list matches TX
  - Initialize locks for object-list to clear
  - Remove object-list

## WAL

The authority writes changes - objects to insert, objects to delete, new certificates, SignedEffects, etc. -- into the WAL rather than a database.   All writes need to be sequenced so it is best to have a single writer.  The writer sequences all writes into a final local order.

The WAL as a write-ahead log has two "watermarks" if you will:

1. HEAD - When the WAL is flushed to disk, that is a commitment that the results of a transaction will be persisted.
2. Persisted - there is a lagging watermark which represents all the items which have been successfully persisted in their final form, whether in a local or remote database.  

In a failure scenario, everything after the Persisted watermark might not have been fully persisted.  Thus, the storage component would have to replay the WAL from the persisted watermark to the HEAD.
The authority should wait for the storage component to be caught up, so that it can start processing with all the state brought up to the HEAD of the WAL.

## Possible Storage Backends

The Storage Component(tm) consists of two lower level APIs:
* An Object storage layer which focuses on storing objects and their indices
   * Provides capabilities to snapshot the current object state
   * Basically we want a distributed, immutable object trie of some kind or object hash map where we can control the versioning
* A Transactions storage layer which focuses on certificates, effects, and old transactions storage and pruning
   * Which can rotate the "hot write bucket" and freeze old ones
   * However the bucket boundaries need to line up with state agreement and epochs

**Comment by velvia** [comment]
After a discussion with @lxfind I wanted to highlight some things regarding the changes this design will bring with regard to atomicity.

- Due to the separation of storage concerns between locking, object storage and transaction storage, we will have to move away from the model of one giant transaction for writing everything at the end of a transaction.
- The above was a bottleneck in any case because locking multiple tables slows down write progress on all of them
- Instead, what we do is, after the object writes succeed, we will then invoke the locking service to do the updates, to create new locks for new objects for example.  The presence of the locks is done after object store succeeds as a guarantee that the objects exist.
- As for when the transactions get written, that will need a different guarantee, but those are not read as frequently.

I'll update the overview with the above details.

^^ @gdanezis

**Comment by gdanezis** [comment]
Ack

>Due to the separation of storage concerns between locking, object storage and transaction storage, we will have to move away from the model of one giant transaction for writing everything at the end of a transaction.

This is in itself a benefit I feel.

**Comment by gdanezis** [comment]
On the basis this is not blocking / breaking other things I put that as pre-mainet
